### PR TITLE
Fix resolving mixed + array

### PR DESCRIPTION
**Pull request by rvanvelzen**
Fixes phpstan/phpstan#7492

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
The failing test is obviously wrong. But we should preserve that for example `int + array` is ErrorType.

File: src/Reflection/InitializerExprTypeResolver.php [hunk 1]
```diff
@@ -924,6 +924,19 @@ public function getPlusType(Expr $left, Expr $right, callable $getTypeCallback):
 			]);
 		}
 
+		if (
+			($leftType->isArray()->yes() || $rightType->isArray()->yes())
+			&& ($leftType instanceof MixedType || $rightType instanceof MixedType)
```
Comment (ondrejmirtes) [inline src/Reflection/InitializerExprTypeResolver.php:929]: My suspicion is that this has a number of problems. Doing `instanceof *Type` is rarely right so it's not hard to poke holes into the logic - this handles the `mixed` scenario but what about something else that "may be" an array or int? What if we do `int|array + int|array`? What piece of code will handle that?

Comment (ondrejmirtes) [reply src/Reflection/InitializerExprTypeResolver.php:929]: Also this condition will catch `TemplateMixedType`, maybe there's a situation which will cause generics to be lost.

Comment (ondrejmirtes) [reply src/Reflection/InitializerExprTypeResolver.php:929]: Also - MixedType is subtractable, meaning there can be a `mixed~array|int` type which means "this is mixed but it's definitely not int nor array". My point is - doing `isSuperTypeOf()` is much safer and usually more correct than instanceof :) More about it here: https://phpstan.org/developing-extensions/type-system#querying-a-specific-type

**Review by staabm — APPROVED** [review]

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Pull request closed without merging**